### PR TITLE
Split out and/or operator handling from infix.

### PR DESCRIPTION
**Pull request by jonmeow**
This is also doing the parse node split, allowing lower reliance in formatter on the tokenized buffer (something that I may be touching more due to import handling).

### CODE REVIEW

File: toolchain/parse/state.def [hunk 1]
```diff
@@ -488,7 +493,7 @@ CARBON_PARSE_STATE(ExprLoop)
 // Completes an ExprLoop pass by adding an infix operator, then goes back
 // to ExprLoop.
 //
-// expr <binary operator> expr ...
+// expr <infix operator> expr ...
 //                            ^
```
Comment (zygoloid) [inline toolchain/parse/state.def]:
```suggestion
//                           ^
```

File: toolchain/parse/state.def [hunk 2]
```diff
@@ -471,7 +471,7 @@ CARBON_PARSE_STATE(ExprInPostfixLoop)
 
 // Handles processing of an expression.
 //
-// expr <binary operator> ...
+// expr <infix operator> ...
 //      ^~~~~~~~~~~~~~~~~
```
Comment (zygoloid) [inline toolchain/parse/state.def]:
```suggestion
//      ^~~~~~~~~~~~~~~~
```

File: toolchain/parse/state.def [hunk 3]
```diff
@@ -480,6 +480,11 @@ CARBON_PARSE_STATE(ExprInPostfixLoop)
 //      ^~~~~~~~~~~~~~~~~~
 //   1. ExprLoop
 //
+// expr <short circuit operator> ...
+//      ^~~~~~~~~~~~~~~~~
```
Comment (zygoloid) [inline toolchain/parse/state.def]:
```suggestion
// expr <short circuit operator> ...
//      ^~~~~~~~~~~~~~~~~~~~~~~~
```

File: toolchain/parse/state.def [hunk 4]
```diff
@@ -501,6 +506,14 @@ CARBON_PARSE_STATE(ExprLoopForBinary)
 //   1. ExprLoop
 CARBON_PARSE_STATE(ExprLoopForPrefix)
 
+// Completes an ExprLoop pass by adding a short circuit operator, then goes back
+// to ExprLoop.
+//
+// expr <short circuit operator> expr ...
+//                              ^
```
Comment (zygoloid) [inline toolchain/parse/state.def]: Isn't this here:
```suggestion
// expr <short circuit operator> expr ...
//                                   ^
```

**Review by zygoloid — APPROVED** [review]
Sure, seems reasonable, and I like that the formatter doesn't need to look at tokens any more.

File: toolchain/parse/state.def (hunk 2 above)
Comment (jonmeow) [reply toolchain/parse/state.def]: Done

File: toolchain/parse/state.def (hunk 3 above)
Comment (jonmeow) [reply toolchain/parse/state.def]: done

File: toolchain/parse/state.def (hunk 1 above)
Comment (jonmeow) [reply toolchain/parse/state.def]: done

File: toolchain/parse/state.def (hunk 4 above)
Comment (jonmeow) [reply toolchain/parse/state.def]: Done

**Pull request merged**